tests/integration/test_upgrade.py

```diff
@@ -16,20 +17,53 @@
 from pytest_operator.plugin import OpsTest
 from tenacity import before_sleep_log, retry, stop_after_attempt, wait_fixed
 
-from .helpers import Bundle, get_leader, get_rsc
+from .helpers import CHARMCRAFT_DIRS, Bundle, get_leader, get_rsc
+
+CHARM_UPGRADE_FROM = "1.32/beta"
+log = logging.getLogger(__name__)
+
+
+def charm_channel_missing(charms: Iterable[str], channel: str) -> Tuple[bool, str]:
+    """Run to test if a given channel has charms for deployment
+
+    Args:
+        charms: The list of charms to check
+        channel: The charm channel to check
+
+    Returns:
+        True if the charm channel or any lower risk exists, False otherwise
+        Returns a string with the reason if True
+    """
+    risk_levels = ["edge", "beta", "candidate", "stable"]
+    track, riskiest = channel.split("/")
+    riskiest_level = risk_levels.index(riskiest)
+    for app in charms:
+        for lookup in risk_levels[riskiest_level:]:
+            out = subprocess.check_output(
+                ["juju", "info", app, "--channel", f"{track}/{lookup}", "--format", "yaml"]
+            )
+            track_map = yaml.safe_load(out).get("channels", {}).get(track, {})
+            if lookup in track_map:
+                log.info("Found %s in %s", app, f"{track}/{lookup}")
+                break
+        else:
+            return True, f"No suitable channel found for {app} in {channel} to upgrade from"
+    return False, ""
```

Comment:
There are cases (like this one) where there's nothing to upgrade FROM.  I had to artificially place something in the 1.32/beta branch just so that tests have something to upgrade from.  

I certainly like this "skipif" feature on the module.  It may make sense in the future to run through multiple upgrades

test-1: upgrade to THESE changes from the last stable release 
test-2: upgrade to THESE changes from the last beta on this channel